### DISCUSSION TITLE
Using SQLCipher version

### DISCUSSION THREAD

**Discussion by chrisjlocke** [sqlitebrowser/sqlitebrowser#2579] [General]
I've not used the SQLCipher version before, but using it to test a .net 'plugin'.
I've created a database, then clicked 'tools' and 'set encryption'.  The dialog appears, when I enter a key.  It's only a small test database, so quick to 'encrypt'.  An open dialog then appears.

![image](https://user-images.githubusercontent.com/7872218/107068321-616c0c80-67d8-11eb-8c77-3dcad20a8e6e.png)

It's not immediately obvious what to do - open '1.db' ?  Tried that, and that is unencrypted.
However, there is a hidden .enctemp file.

![image](https://user-images.githubusercontent.com/7872218/107068426-82346200-67d8-11eb-870e-d5f8bec4ced3.png)

This *does* seem encrypted.  Is this how it works - I now use the '.enctemp' file and rename the '1.db' as no longer being used?
Just seems that there should be some dialog explaining this, or even a 'do you want the original file backed up and the '.enctemp' renamed?' helpful last step...
Initially I just thought the encryption wasn't working...

Also, in the encryption dialog, you can set v3 or v4 defaults.

![image](https://user-images.githubusercontent.com/7872218/107068869-0c7cc600-67d9-11eb-81ae-490777793438.png)

Does this also set whether a v3 or v4 type file is created?  I seem to remember something hazy about the file types are completely different, and I wasn't sure what version my .net code can open - can I set which version DB4S creates?  Or does it still create a v4 file, just with v3 'headers' / settings?

**Comment by justinclift** [comment]
Hmmm.  For me, when I create a database file using the SQLCipher version (it's v4 by default btw), it "just works".  Maybe you need to press the "Write Changes" button before it commits the data though?  (Hmmm, I don't think that's actually the case, but I'm not 100%).

That `.enctemp` file is likely to be the encrypted version of the database, I'm not sure why it hasn't been swapped into place of the .db file though by the time you're looking.

**Comment by MKleusberg** [comment]
The encryption process create a new encrypted database file  (the .enctemp file), copies all the data to the new file, deletes the original file, and then renames the .enctemp file to replace the original file. Seems like in your case closing or deleting the original file didn't work. As a first guess: Is it opened by another application at the same time?

You could say the v3/v4 file formats are the same. The only difference are the crypto settings they use. But since these are easy to change, you can open a v3 file in SQLCipher v4 and vice-versa - you only need to set the parameters before doing so.

**Comment by chrisjlocke** [comment]
Starting the SqlCipher version from scratch and creating a new database still fails.

https://user-images.githubusercontent.com/7872218/107280059-bebdc300-6a4f-11eb-9f07-23caa05b3005.mp4

Bizarre.  Will keep tinkering.  Might be something weird like new files are scanned by anti-virus or something (this is on a work PC to which I don't have much control - will try on my other laptop that doesn't have anything on it...)

**Comment by justinclift** [comment]
Tried out the Win64 nightly build yesterday, and the same thing happens.

The database file itself after the encryption process is the un-encrypted one (which leads DB4S to prompt for which file to open), and there is an .enctemp file next to it.  Opening the .enctemp file works, as that's the encrypted SQLCipher one.

Haven't looked into it further at this point, but figured I'd at least mention here that it's not just a @chrisjlocke thing. :smile:

**Reply by chrisjlocke** [reply]
>it's not just a @chrisjlocke thing.

🤨
😁

**Comment by chrisjlocke** [comment]
I had a look at 'fileClose' but couldn't spot anything obvious, but wondering if that's resetting 'db.currentFile' ?

    fileClose();
    QFile::rename(db.currentFile(), db.currentFile() + ".enctempold");

So the file is closed, then it's trying to be renamed.  Either that, or it's happening too quick - closing but the file isn't closed yet, so the rename is failing.  I see some 'qApp->processEvents();' in the lines above...
The file is being closed, but the rename isn't happening (and nothing else past that point...) 
The next line:
    if(fileOpen(db.currentFile()))
Wonder if that's why it opens a dialog ... blank filename?
I'd add a line to throw out to the application log but closing the database also clears the log (doh) and writing to the console doesn't work in Windows - running it from the command-line just spawns DB4S and returns to the prompt - doesn't 'listen' for output...

![image](https://user-images.githubusercontent.com/7872218/111919769-9039fb80-8a83-11eb-9cad-e578147d4534.png)

**Reply by chrisjlocke** [reply]
Interestingly, clicking 'File' and 'Save as' also fails in sqlCipher version...

![image](https://user-images.githubusercontent.com/7872218/113429094-dec68e80-93cf-11eb-8b35-4385bf912653.png)

A 0 byte file is created.
A non-encrypted database saves fine (no errors).

/raises eyebrow in a puzzled fashion.

**Reply by mgrojo** [reply]
Does it print anything in the 'Error log' inside the 'SQL Log' dock (yes, it is no longer only SQL, by the way)?

**Reply by chrisjlocke** [reply]
Unfortunately not - I did put above, "I'd add a line to throw out to the application log but closing the database also clears the log (doh) and writing to the console doesn't work in Windows - running it from the command-line just spawns DB4S and returns to the prompt - doesn't 'listen' for output..."

As soon as you tell it to encrypt, set a password and click OK, the 'Open database' dialog is displayed, so the log is cleared.

Edit: Whoopsie!  Sorry!  The application log is cleared, but the error log isn't.  Yes, there is an error there:
`(1) QFile::rename: Empty or null file name (, :0)`

So my theory might be right then - closing the database resets the variable for the next line.... ?
